### PR TITLE
Extend concrete evaluation

### PR DESCRIPTION
**Pull request by joaomhmpereira**
to address https://github.com/formalsec/encoding/issues/99

### CODE REVIEW

**Comment by filipeom** [comment]
Hey João, sorry for the massive delay. However, in an effort to reduce your workload, I've updated issue #99. I believe we've completed most of the operators, but please pay special attention to the ones annotated with :exclamation:, as I'm not sure if we've included them yet.

In this PR, you can remove the float operators you added. I've begun moving them directly to ecma-sl here: [formalsec/ECMA-SL#stdlib](https://github.com/formalsec/ECMA-SL/tree/stdlib)

Regarding the string operators you included, I'd like to keep them, excluding the lowercase/uppercase functions. We'll also transfer those to ecma-sl's standard library.

File: lib/value.ml [hunk 1]
```diff
@@ -43,6 +47,8 @@ let pp fmt (v : t) =
   | Real x -> fprintf fmt "%F" x
   | Num x -> Num.pp fmt x
   | Str x -> Format.fprintf fmt "%S" x
+  | List _ -> assert false
+  | Tuple _ -> assert false
```
Comment (filipeom) [inline lib/value.ml:51]:
```suggestion
  | List l -> 
    fprintf fmt "[%a]" 
      (pp_print_list ~pp_sep:(fun fmt () -> pp_print_string fmt ", ")
      pp)
      l
  | Tuple _ -> same but with "(%a)"
```

You would need to add `rec` to the definition as well

**Review by filipeom — COMMENTED** [review]
By the way, could you add simple tests to `test_{unop|binop|triop|cvtop}`, similar to the ones already present, just to verify if the construction of the operators you added is being correctly simplified in practice?

I don't want to rush this. I'm going to pin this branch to https://github.com/formalsec/ECMA-SL/tree/stdlib and see if we can start the value migration

**Comment by joaomhmpereira** [comment]
@filipeom I think this is finally ready for review

File: lib/ty.ml [hunk 2]
```diff
@@ -124,6 +131,12 @@ type cvtop =
   | String_from_code
   | String_to_int
   | String_from_int
+  | String_to_float
+
+type naryop =
+  | AndN
+  | OrN
```
Comment (filipeom) [inline lib/ty.ml:138]: These can only be used with logical operators so we can probably rename them to:

```suggestion
  | Logand
  | Logor
```

File: lib/ty.ml [hunk 3]
```diff
@@ -72,13 +76,15 @@ type binop =
...
-  | Seq_concat
-  | Seq_prefix
-  | Seq_suffix
-  | Seq_contains
-  | Seq_last_index
+  | String_prefix
+  | String_suffix
+  | String_contains
+  | String_last_index
+  | String_split
```
Comment (filipeom) [inline lib/ty.ml:87]:
```suggestion
```

I didn't see a String_split in `Eval`. Maybe we can remove this then?

**Comment by filipeom** [comment]
Sorry for the delay, I did a quick check of the code, I'll try to give more feedback as I start to get more time

**Comment by joaomhmpereira** [comment]
> There are a few more changes I want to add here before we merge. @joaomhmpereira you merged main into here on your last commit [fe136d2](https://github.com/formalsec/smtml/pull/101/commits/fe136d27c5e91545ed2b2cfbab51e49a2b1e4949)?

Yes, but it was an older version where the tests for the solver were not fixed yet

**Comment by filipeom** [comment]
> Yes, but it was an older version where the tests for the solvers were not fixed yet

Okok, so I'll rebase it on main to fix conflicts as well

**Comment by filipeom** [comment]
Don't know why the CI is failing on bitwuzla. But it should be ok

**Pull request merged**